src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java

```diff
@@ -279,11 +279,30 @@ public void onSubscribe(Flow.Subscription subscription) {
             subscription.request(1);
         }
 
+        // Arbitrary value chosen to tolerate short writes, while avoiding
+        // potential infinite busy loop.
+        private static final int MAX_WRITE_RETRIES = 10;
+
         @Override
         public void onNext(List<ByteBuffer> items) {
             try {
-                out.write(items.toArray(Utils.EMPTY_BB_ARRAY));
-            } catch (IOException ex) {
+                long expectedSize = Utils.remaining(items);
+                long bytesWritten = 0;
+                ByteBuffer[] buffers = items.toArray(Utils.EMPTY_BB_ARRAY);
+                int retries = 0;
+                while (bytesWritten < expectedSize) {
+                    long n = out.write(buffers);
+                    if (n <= 0) {
+                        retries++;
+                        if (retries > MAX_WRITE_RETRIES) {
+                            throw new IOException("zero bytes written");
+                        }
+                    } else {
+                        retries = 0;
+                        bytesWritten = Math.addExact(bytesWritten, n);
+                    }
+                }
+            } catch (Exception ex) {
```

Comment:
I believe we should change that line to `catch (IOException | ArithmeicException) {`

Reply:
Agreed. Testing it locally before pushing.

Reply:
Alternatively it seems that moving the try down to line 292 works too.